### ISSUE TITLE
Store blob archives more efficient

### ISSUE THREAD

**Issue by tsteur** [piwik/piwik#7373]
Current solution is to store each datatable in one archive_blob, and each subtable in a separate archive_blob. If a datatable has 1000 subtables (which is not unusual), we create 1001 rows in the archive table for this single report.

New solution would be to store max 2 blobs per archive: One for the first level datatable, and another one combined for all subtables. We do not store them combined in one blob entry to have still max performance for first level datatables while loading a subtable will be slightly slower since we will have to unserialize all rows. Storing all subtables in one row will not only result in less entries in the blob table and therefore a lower index size, it will also result in a lower data size (less storage needed) as GZIP can compress much better. For example currently when storing only one row per subtable, we do store `Piwik\DataTable\TableViewRow` 1000 times whereas afterwards it will be only once since gzip does not deal with 1000 different compressed entries but one.

Storing the blob in one main blob and one for subtables will also allow us at some point for example to allow users to delete all archives for subtables but not for the first level tables.

Here are some results of a comparsion for one site. I did track about 100.000 pageviews on one day and archive it afterwards with old behavior and with new, suggested behavior:

### SINGLE SUBTABLE ENTRIES (OLD BEHAVIOUR)
0 visits in last last2 days, 0 visits today, Time elapsed: 0.708s
14210 visits in last last2 weeks, 0 visits this week, Time elapsed: 57.050s
14210 visits in last last2 months, 0 visits this month, Time elapsed: 7.707s
14210 visits in last last2 years, 14210 visits this year, Time elapsed: 0.694s
Time elapsed: 66.168s [1/1 done]

#### Archive 2015_02 
| Rows in blob table  | Index size | Data size | 
| ------------- |  -------------  | ------------- |
| 4.731  |  400K | 5.5MB |

### WITH SUBTABLES GROUPED (NEW BEHAVIOUR)

0 visits in last last2 days, 0 visits today, Time elapsed: 0.831s
14210 visits in last last2 weeks, 0 visits this week, Time elapsed: 61.312s
14210 visits in last last2 months, 0 visits this month, Time elapsed: 9.312s
14210 visits in last last2 years, 14210 visits this year, Time elapsed: 6.646s
Time elapsed: 78.122s [1/1 done]

#### Archive 2015_02 
| Rows in blob table  | Index size | Data size | 
| ------------- | -------------  | ------------- |
| 165  | 16K | 3.5MB |

### Archives comparison
* 30 times less rows!
* Index size is 30 times smaller! (400K vs 16KB)
* Data size is halved (5.5MB vs 3.5MB)
* Archiving is a bit slower with less data but probably 5% faster the more archives there are in there (could be also by accident, don't think there will be huge improvement. While we have to do less index work and row writing we do have a bit more serialize work todo)

The advantage will be even more clear when also having more other reports with subtables. For example there were no events tracked, no content tracking. So the factor can grow from 30 times to 50 times. On the other side if one has only a few subtables the advantage will be less clear.

### Reading comparison
* Building expanded datatable (25k rows, 800 subtables)
 * Speed pretty much the same (about 500ms with xhprof on my local instance)
 * Same memory usage
* Building 1st level datatable: 
 * Speed is the same (about 10ms with xhprof)
 * Same memory usage
* Building a subtable datatable: 
 * Speed is slower (12ms vs 40ms) as there is more data to uncompress (0.175ms vs 26.970ms) and to unserialize (0 vs 3.561ms). Background: we have to unserialize the blob that contains all subtables first before we can access the serialized subtable that we are interested in.
 * Peak memory usage 13MB vs 28MB (new method needs more memory since we have to unserialize all blobs into one array which contains the `[idSubtable] => [serialized subtable rows]`)
 * Incl. memory usage is always 13.2MB

Expanded and first level reports are same in time needed to load archive and create datatable. Subtable reports are slightly slower and have a higher peak memory. Once there are more and more archives in a blob table the new method should become a bit faster as the index will be smaller in Mysql, there will be less data to transfer from mysql to php, there will be less archives to parse in PHP, ...

Eg when having on a test server 2GB of blob entries, 1.6M rows and 300MB index size could go down to 1GB storage, 50K rows and 10MB index size.

**Comment by diosmosis** [comment]
To speed up single subtable reading, maybe you could grep through the serialized string for the [idSubtable] indexes? Maybe there could be a special value used to look for in the serialized string (ie, something that won't normally occur).

**Comment by tsteur** [comment]
> Maybe there could be a special value used to look for in the serialized string (ie, something that won't normally occur).

Thought about this already but people could use any string on purpose in tracking requests, eg pageviews and break any archive?

> To speed up single subtable reading, maybe you could grep through the serialized string for the [idSubtable] indexes? 

That could be possible but probably could have same issue as above?

Background: To serialize subtables into one blob I currently do the following

```php
$combinedSubtableBlob = serialize(array(1 => 'serialized subtable', 2 => 'serialized subtable3', $subtableId => $serializedSubtable....));
```

Then when reading the subtable archive:
```php
$subtables = unserialize(combinedSubtableBlob); // array that holds all serialized subtables
```

I'm not really concerned about those 20ms on top as it will be faster on a fast CPU, and it will be faster when having less subtables (this test had 800 subtables which is often less) but it could be as well more when having eg 1600 subtables. Would be nice to make it faster though! Hope to test it on a fast server next week

**Comment by diosmosis** [comment]
> Thought about this already but people could use any string on purpose in tracking requests, eg pageviews and break any archive?

I used double null chars for something like this before, if it's not possible for users to track null chars, then this could work; and if it doesn't and unserializing fails, we could fallback to the other strategy. Maybe this would work?

**Comment by tsteur** [comment]
I think one can still send null chars via `%00`. I need to test it but pretty sure people could misuse it.

**Comment by mattab** [comment]
I'm not sure about this but here is an idea while I think about it: in case there are performance problems when storing all hundreds sub-tables in this one blob, instead maybe we could store sub-tables by chunks of 100 in each blob. Then given the idSubtable we could still find out which "100-chunk blob" this sub-table is stored in. This would not speed up `expanded=1` or `flat=1` but it could speed up other cases like `recursive search` or `click on a Page URL folder` (or any API request that passe an `idSubtable`)

**Comment by tsteur** [comment]
You mentioned this already and I wouldn't go for it right now. It would be kinda easy to add afterwards. First, as mentioned, we need to test on a big instance with fast server anyway. Otherwise I'm not worried about those 20ms that it adds on my server as said. We still spend much more time in other parts on PHP side that are way worse. Eg `replaceColumnNames` can easily take > 1 seconds. Also on the client side it spends 350ms on my fast Mac with the latest Chrome which can be up to a second or way more (on old browsers) in JavaScript + on top painting etc just to render a datatable.

**Comment by tsteur** [comment]
Also that this only works when using `int subtableId` and no longer once we replace this with proper labels.

**Comment by diosmosis** [comment]
Would users be interested in converting old archive tables to this approach (to save space)? If so, we could add a command to do so in another pull request.

**Comment by diosmosis** [comment]
It seems odd (and potentially problematic) that the archiving time increased by ~18%. Can you replicate on subsequent runs? Ie, if you run archiving 5 times on old code and average the time (before each time dropping tables & maybe restarting the mysql server), then run archiving 5 times on new code & average the time, is the time difference still that large? If so, some more profiling may be a good idea.

**Comment by tsteur** [comment]
FYI: I wanted to test how it performs when writing 200 subtables into one blob each. While doing this I noticed some blobs include only 3 subtables instead of 200 since those subtableIds are not consecutive see eg here: http://demo.piwik.org/index.php?module=API&method=API.getProcessedReport&apiModule=Actions&apiAction=getPageUrls&idSite=3&period=day&date=2015-03-18&format=xml&token_auth=anonymous

There is a subtableId `21` and `26`. Other ids seem not to be used. 

I'll have to think of another algorithm to group those tables into chunks of 100 or 200 or ...

**Comment by tsteur** [comment]
I made some progress on this here but I am not finished yet.

### Subtable chunks
First regarding splitting the subtables into chunks. I did a test with storing up to 200 DataTables into one blob. Serialized and gzip compressed this takes the following amount of space in one row:

* 100 rows in each subtable: 0.5MB
* 1000 rows in each subtable: 5MB
* 20 rows in each subtable: 86KB

I think for most instances the regular case is having on average 20 rows in each subtable or even less. Since there are often less leaves in subtables (for example on Piwik under `/2014` there are only 12 leaves etc). The worst case for most action and referrer reports are by default 100 rows per subtable as configured in `global.ini.php`. This would take only 0.5MB and would not be a problem. Problem are for example Goal/Ecommerce reports where there is no limit for subtables so we have to assume that we could have 200 subtables with for example 1000 rows each on average. This would take 5MB and is maybe too much. In reality we would have maybe a few subtables with 1000 rows or 5000 rows and many with much less but we should take the worst case since the number of subtables to store in one blob will be hard to change later. Therefore, I'd probably go with only 100 subtables per blob. This results in almost twice the amount of blob entries in the archive tables and almost twice the index size (compared to having 200 subtables in one blob) but still up to 100 times better than before.

I did a test on our big testing instance where I archived one day. In that new branch with 100 subtables per blob archiving time dropped from `159 seconds` and `118MB memory` to `109 seconds` and `94MB`. Main improvement was PDO time that dropped from `99 seconds` to `56 seconds` as less inserts to write etc. We also gain another `7 seconds` due to serialize improvements (only serializing arrays, faster Row, more about this further down) even though we have to spend more time when serializing 100 subtables into one blob each time etc.

### Serializing only arrays of all rows instead of `Row` instances

I think it does make sense to store only a serialized array in the archive blob in the future and not the serialized `Row` instance. This is easy to implement as it was kinda already prepared [here](https://github.com/piwik/piwik/blob/2.12.0-rc3/core/DataTable.php#L1278). To make it short: Serializing/ unserializing will be faster + take less memory, and we will need a bit less storage in the DB but those are not really the primary reasons for it. It will give us more flexibility on refactoring Row instances etc. For example we will be no longer dependent on `\Piwik\DataTable\Row` but can easily refactor things. Keeping BC for old archives will be no problem and won't add much overhead.

#### What to serialize?

We would serialize that array that is usually passed to `new Row($row)`. So that means `array(Row::COLUMS => array(...), Row::METADATA => array(...), Row::DATATABLE_ASSOCIATED => int)`. So to be more concrete: `array(0 => array(...), 1 => array(...), 3 => 5)`. In my current implementation I do for example always archive an entry such as the associated subTableId `Row::DATATABLE_ASSOCIATED => null` even when no value is set. We could save more storage if we do not serialize an associated DataTable or metadata key if they are not set. So sometimes we could only archive `array(ROW::COLUMNS => array(...))` but I'm not sure if it is good.

#### Archive storage needed when archiving

| Method        | Serialized MB           | Compressed MB  |
| ------------- |:-------------:| -----:|
| arrays      | 9.7MB (9756235 bytes) | 1.1MB (1103768 bytes) |
| Row instances      | 11.85MB (11856102 bytes)      |   1.14MB (1141680) |

There will be only a very minor difference in storage size but we will require less DB storage. We won't save much storage since gzip works pretty well :)

#### Performance unserializing 200 DataTables with 50k rows each

| Method        | Time in ms           | Memory
| ------------- |:-------------:| -----:|
| Unserializing array and creating Row instance      | 25627 (6000ms of this time are spent in creating Row instances) |  7.6MB | 
| Unserializing Row instances      | 31230    |  8.12MB | 

Unserializing is 6 seconds faster + we need less memory! We could even further improve the time needed to create the arrays by having a special Row instance that does not check whether eg `$row[0], $row[1] and $row[3]` are set in the [constructor](https://github.com/piwik/piwik/blob/2.12.0-rc3/core/DataTable/Row.php#L80) since we will know they are there, unless we do not archive them see previous topic "what to serialize". 

#### Performance serializing 200 datatables with - I think - 100 rows each

| Method        | Time in ms           | 
| ------------- |:-------------:| -----:|
| array      | 20318 |  7.6MB |
| Row instances      | 34772    |  8.12MB |

Serializing an array is 14 seconds faster + we need less memory!

### Faster `DataTable\Row`

Those changes are kinda unrelated to this PR and we can merge this separately but it is kinda related since we maybe only want to serialize the arrays etc. The performance improvements will make archiving much faster but especially Range archives and aggregated reports etc. Normal reports will be as well much faster but maybe a bit less.

#### Faster columns access
I started working on a faster `DataTable\Row` implementation since we spent in `getColumn()` 33% of our time when archiving a big year instance etc. Currently everything is stored in `$c = array(0 => columns, 1 => metadata, 3 => subtableId)`. Whenever we want to get one column we have to do `return $this->c[self::COLUNNS][$column]` which is twice as slow as `return $this->columns[$column]`. After changing this `getColumn` dropped from 33% to 17% (from 900 seconds to 450 seconds). 

#### Faster subtables access
We iterate over all subtables very! often on big archives etc. `getSubtable` and `isSubtableLoaded` take about 10% of the total time to archive that year. This will be afterwards improved to 0.1%. What made it slow is having to copy many integers and doing many checks with negative subtableIds and such things: `return abs($this->c[self::DATATABLE_ASSOCIATED])`. Instead we will maintain a bool property that defines whether a subtable is loaded or not and a separate property for the subtableId.

#### Further ideas

I'm playing around with a `Row` class that is based on `\ArrayObject` which seems to make `getColumns`, `setColumn`, ... again 3 times faster. Since we still spend 20% of the time there, this is a huge thing. We could sometimes even further improve that time by not using methods like `getColumns` (in `SumRow`, ...) etc but that is not really important yet. Problem: It seems to take more memory. I had only 25k row instances but memory increased from 104MB to 108MB. When dealing with 500K row instances which I do when archiving a year on my test instance this can become a problem but needs to be checked.
 
### A real life comparisons
In a real life flatten report with 25k rows I noticed a drop from `5.6 seconds` to `4.1 seconds` and memory dropped from `126MB` to `104MB` with this change.

On our big testing instance I can reproduce same behaviour where execution time dropped from `8.0 seconds` to `6.4 seconds` and memory dropped from `100MB` to `85MB` for a bigger flatten report.

**Comment by tsteur** [comment]
Here's another update re faster Row implementation and the other changes I made so far. By letting Row inherit `ArrayObject` the time spent in `getColumn` goes down from `33%` to `0.3%`. This is a 900 seconds improvement out of a 2900 seconds total request time! 

Here some numbers from a test with xhprof (xdebug disabled of course) when archiving a year that has quite some data.

| Method        | Before           | After  |
| ------------- |:-------------:| -----:|
| Total request time      | 2970s | 1750s (actually about 1200 seconds, more about it later) |
| Total incl. memory      | 210MB      |   210MB |
| Peak memory usage      | 4506MB      |   3656MB |

The total request time improved to 1750 seconds which is actually about 1200 seconds. Why? About 425 seconds were spent in `__wakeup` which is just there for being BC with old archives that serialize the whole `Row` instance. This code will not run for new archives. Once we no longer have to run the `__wakeup` we might even save another 1 GB of memory. On top I have more improvements in the pipe line in other PRs that will improve the request time by another > 100 seconds. This is a 60% performance boost.
